eduNEXT/edx-platform
# feat: add grading method support

## Description
This PR adds a new field in the Problem settings for choosing a **Grading Method**. Currently, the only Grading Method is the Last Score. From now on, the new grading methods allow are:

1. **Last Score _(Default)_**: The last score made is taken for grading.
2. **First Score**: The first score made is taken for grading.
3. **Highest Score**: The highest score made is taken for grading.
4. **Average Score**: The average of all scores made is taken for grading.

## Dependencies
This PR works with the legacy Problem Editor interface, but if you use the new interface with course-authoring MFE, you need the changes in [this PR](https://github.com/eduNEXT/frontend-lib-content-components/pull/2).

### Legacy Interface
![image](https://github.com/eduNEXT/edx-platform/assets/64033729/7a1bd0a8-f749-464c-a436-c5993ff73e24)
### MFE Interface
![image](https://github.com/eduNEXT/edx-platform/assets/64033729/910bbaba-ab07-433f-9314-d5c9b299b120)

## Testing Instructions
1. Go to Studio and create a problem component from a Unit.
2. In component settings you should see a new field: `Grading Method`. This field is a dropdown list with the Grading Methods mentioned above.
3. Choose a Grading Method and save changes.
4. From the LMS answer the problems and check with the different Grading Methods. You can also check from the progress section.

### Rescoring
1. Go to Studio and create a problem component from a Unit. e.g. `Advanced` > `Blank Problem`. In the settings you can add this example of a problem:

```xml
<problem>
  <h3 class="hd hd-2 problem-header">Multiple Choice</h3>
  <multiplechoiceresponse>
    <choicegroup type="MultipleChoice">
      <choice correct="true">correct</choice>
      <choice correct="false">incorrect 1</choice>
      <choice correct="false">incorrect 2</choice>
    </choicegroup>
  </multiplechoiceresponse>
  
  <h3 class="hd hd-2 problem-header">CheckBox</h3>
  <choiceresponse>
    <checkboxgroup>
      <choice correct="true">correct 1</choice>
      <choice correct="true">correct 2</choice>
      <choice correct="false">incorrect</choice>
    </checkboxgroup>
  </choiceresponse>
  
  <h3 class="hd hd-2 problem-header">Drop-Down List</h3>
  <optionresponse>
    <optioninput options="('resp 1', 'resp 2', 'resp 3')" correct="resp 3"></optioninput>
  </optionresponse>
</problem>
``` 
2. In component settings choose a grading method, e.g. Highest score.
3. From the LMS answer the problems and check your score.
4. From Studio edit the correct answers to the problem.
5. From the LMS as an instructor, make the rescore of the problem in `STAFF DEBUG INFO` > ` Rescore Learner's Submission` 
6. The final score should have changed depending on the answers.
